include/zephyr/devicetree/mem.h

```diff
@@ -0,0 +1,107 @@
+/**
+ * @file
+ * @brief Memory Devicetree macro public API header file.
+ */
+
+/*
+ * Copyright (c) 2023 Graphcore Ltd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DEVICETREE_MEM_H_
+#define ZEPHYR_INCLUDE_DEVICETREE_MEM_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <stdint.h>
+#include <string.h>
+#include <stdbool.h>
+
+#include <zephyr/devicetree/memory-attr.h>
+
+/**
+ * @defgroup devicetree-memory Devicetree memory API
+ * @ingroup devicetree
+ * @{
+ */
+
+/**
+ * @brief Container for memory-region information specified in devicetree
+ *
+ * This type contains pointers to the begin and end of the memory region and
+ * a string that specifies such region's properties.
+ *
+ */
+struct dt_mem_region {
+	const uintptr_t start;
+	const uintptr_t end;
+	bool isnocache;
+};
+
+/** @cond INTERNAL_HIDDEN */
+
+#define _DT_IS_NOCACHE_MEM_REGION(node_id) \
+	COND_CODE_1(DT_ENUM_HAS_VALUE(node_id, zephyr_memory_attr, RAM_NOCACHE),    \
+			(true),                                                             \
+			(false))
+
+/** @endcond */
+
+/**
+ * @brief Can be used to initialize a @see{struct dt_mem_region}.
+ *
+ * @param node_id Node ID of the memory region declared in device tree from
+ * which the information will be obtained.
+ *
+ * @example
+ *
+ *		struct dt_mem_region mem_r = {
+ *			DT_GET_MEM_REGION(DT_NODELABEL(<node_label>))
+ *		};
+ *
+ */
+#define DT_GET_MEM_REGION(node_id)                                              \
+	{                                                                           \
+		.start = (const uintptr_t)DT_REG_ADDR(node_id),                         \
+		.end =                                                                  \
+			(const uintptr_t)(DT_REG_ADDR(node_id) + DT_REG_SIZE(node_id)) - 1, \
+		.isnocache = _DT_IS_NOCACHE_MEM_REGION(node_id)                         \
+	},
+
+/**
+ * @brief Return true if @p buf is in a nocache memory region.
+ *
+ * @param buf Buffer to be checked
+ * @param len_bytes Length of the buffer above
+ *
+ */
+static inline bool dt_buf_in_nocache(uintptr_t buf, size_t len_bytes)
+{
+	const struct dt_mem_region mem_regions[] = {
+		DT_MEMORY_ATTR_FOREACH_NODE(DT_GET_MEM_REGION)
+	};
```

Comment:
hard no. You are actually allocating RODATA by only including an header file whether you use it or not.

Reply:
Again, this was in the SPI driver source file only until I got requested to move to a generic place.

Reply:
well, an header is definitely not a proper place for this.